### ISSUE TITLE
Occasional segfault in dynamic batch scheduler using ONNX runtime

### ISSUE THREAD

**Issue by OvervCW** [triton-inference-server/server#4925]
**Description**
We are repeatedly loading ONNX models and performing inference on them. Sometimes the first inference request on a newly loaded model will crash Triton with a segmentation fault.

It is reproducible in both release and debug builds and results in the following stack trace:

```
Thread 23 "tritonserver" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff9affd000 (LWP 2125)]
std::__uniq_ptr_impl<std::mutex, std::default_delete<std::mutex> >::_M_ptr (this=0x70) at /usr/include/c++/9/bits/unique_ptr.h:154
154	      pointer    _M_ptr() const { return std::get<0>(_M_t); }
(gdb) bt
#0  std::__uniq_ptr_impl<std::mutex, std::default_delete<std::mutex> >::_M_ptr (this=0x70) at /usr/include/c++/9/bits/unique_ptr.h:154
#1  0x00007ffff739472e in std::unique_ptr<std::mutex, std::default_delete<std::mutex> >::get (this=0x70) at /usr/include/c++/9/bits/unique_ptr.h:361
#2  0x00007ffff7393eec in triton::core::Payload::GetExecMutex (this=0x0) at /tmp/tritonbuild/tritonserver/build/_deps/repo-core-src/src/payload.h:57
#3  0x00007ffff73901aa in triton::core::DynamicBatchScheduler::Enqueue (this=0x7fff6c07f080, request=std::unique_ptr<triton::core::InferenceRequest> = {...}) at /tmp/tritonbuild/tritonserver/build/_deps/repo-core-src/src/dynamic_batch_scheduler.cc:252
#4  0x00007ffff73b74de in triton::core::Model::Enqueue (this=0x7fff6c0a24e0, request=std::unique_ptr<triton::core::InferenceRequest> = {...}) at /tmp/tritonbuild/tritonserver/build/_deps/repo-core-src/src/model.h:97
#5  0x00007ffff73ae11e in triton::core::InferenceRequest::Run (request=std::unique_ptr<triton::core::InferenceRequest> = {...}) at /tmp/tritonbuild/tritonserver/build/_deps/repo-core-src/src/infer_request.cc:245
#6  0x00007ffff7502cce in triton::core::InferenceServer::InferAsync (this=0x5555565bdec0, request=std::unique_ptr<triton::core::InferenceRequest> = {...}) at /tmp/tritonbuild/tritonserver/build/_deps/repo-core-src/src/server.cc:507
#7  0x00007ffff751d0e7 in TRITONSERVER_ServerInferAsync (server=0x5555565bdec0, inference_request=0x7fff8c41cd30, trace=0x0) at /tmp/tritonbuild/tritonserver/build/_deps/repo-core-src/src/tritonserver.cc:2853
#8  0x000055555564d140 in triton::server::(anonymous namespace)::ModelInferHandler::Process (this=0x555556b35fc0, state=0x7fff8c036500, rpc_ok=true) at /workspace/src/grpc_server.cc:3706
#9  0x000055555565ce14 in triton::server::(anonymous namespace)::InferHandler<inference::GRPCInferenceService::WithAsyncMethod_ServerLive<inference::GRPCInferenceService::WithAsyncMethod_ServerReady<inference::GRPCInferenceService::WithAsyncMethod_ModelReady<inference::GRPCInferenceService::WithAsyncMethod_ServerMetadata<inference::GRPCInferenceService::WithAsyncMethod_ModelMetadata<inference::GRPCInferenceService::WithAsyncMethod_ModelInfer<inference::GRPCInferenceService::WithAsyncMethod_ModelStreamInfer<inference::GRPCInferenceService::WithAsyncMethod_ModelConfig<inference::GRPCInferenceService::WithAsyncMethod_ModelStatistics<inference::GRPCInferenceService::WithAsyncMethod_RepositoryIndex<inference::GRPCInferenceService::WithAsyncMethod_RepositoryModelLoad<inference::GRPCInferenceService::WithAsyncMethod_RepositoryModelUnload<inference::GRPCInferenceService::WithAsyncMethod_SystemSharedMemoryStatus<inference::GRPCInferenceService::WithAsyncMethod_SystemSharedMemoryRegister<inference::GRPCInferenceService::WithAsyncMethod_SystemSharedMemoryUnregister<inference::GRPCInferenceService::WithAsyncMethod_CudaSharedMemoryStatus<inference::GRPCInferenceService::WithAsyncMethod_CudaSharedMemoryRegister<inference::GRPCInferenceService::WithAsyncMethod_CudaSharedMemoryUnregister<inference::GRPCInferenceService::WithAsyncMethod_TraceSetting<inference::GRPCInferenceService::Service> > > > > > > > > > > > > > > > > > >, grpc::ServerAsyncResponseWriter<inference::ModelInferResponse>, inference::ModelInferRequest, inference::ModelInferResponse>::<lambda()>::operator()(void) const (this=0x555556b35fc0) at /workspace/src/grpc_server.cc:2503
...
```

**Triton Information**
v2.25.0

**Are you using the Triton container or did you build it yourself?**

The problem is reproducible with both `nvcr.io/nvidia/tritonserver:22.08-py3` and an image we built ourselves from the v2.25.0 branch using the following command:

```shell
python build.py --build-type=Debug --backend onnxruntime --enable-logging --enable-stats --enable-metrics --enable-gpu-metrics --enable-tracing --enable-nvtx --enable-gpu --endpoint grpc
```

**To Reproduce**
Repeatedly load/unload an ONNX model and send an inference request.

Batching is enabled on all of our models using the `--backend-config=onnxruntime,default-max-batch-size=128` commandline flag (even the ones that technically aren't really meant to be used for batching). With that said, this happens even when sending only a single inference request at a time (with no batch), so I don't think the batch scheduler is actually creating any batches in our scenario.

I think dynamic batching is enabled because autocomplete seems to do it by default using the aforementioned commandline flag.

**Expected behavior**
No crash.

**Comment by OvervCW** [comment]
Follow-up question: is there a way to disable the dynamic batch scheduler and only allow explicit batches when using config auto-complete?

**Comment by rmccorm4** [comment]
Hi @OvervCW ,

Thanks for filing the bug.

> #2  0x00007ffff7393eec in triton::core::Payload::GetExecMutex (this=0x0) at /tmp/tritonbuild/tritonserver/build/_deps/repo-core-src/src/payload.h:57

Looks like the Payload may be `nullptr` ? CC @tanmayv25 @dyastremsky for follow-up / investigation as I'll be on vacation.

> Follow-up question: is there a way to disable the dynamic batch scheduler and only allow explicit batches when using config auto-complete?

Auto-complete should only fill-in what's missing, so if you specify the desired fields related to batch size, dynamic_batching, etc. in your config file - auto complete should respect those fields.

**Comment by OvervCW** [comment]
> Auto-complete should only fill-in what's missing, so if you specify the desired fields related to batch size, dynamic_batching, etc. in your config file - auto complete should respect those fields.

It looks like the ONNX backend always chooses the dynamic batch scheduler when autocompleting a config if no scheduler is explicitly specified, so there is no way to use the default scheduler. I've created [a separate issue](https://github.com/triton-inference-server/onnxruntime_backend/issues/147) for that.

> Looks like the Payload may be `nullptr` ? CC @tanmayv25 @dyastremsky for follow-up / investigation as I'll be on vacation.

I found a race condition in the dynamic batch scheduler that causes this problem and I've soon submit [a PR](https://github.com/triton-inference-server/core/pull/122) to fix it.
